src/target/riscv/riscv-011.c

```diff
@@ -1368,16 +1372,32 @@ static int get_register(struct target *target, riscv_reg_t *value,
 			return result;
 	}
 
-	if (regid == GDB_REGNO_MSTATUS)
-		target->reg_cache->reg_list[regid].valid = true;
-
 	return ERROR_OK;
 }
 
+/* This function is intended to handle accesses to registers through register
+ * cache. */
 static int set_register(struct target *target, enum gdb_regno regid,
 		riscv_reg_t value)
 {
-	return register_write(target, regid, value);
+	assert(target->reg_cache);
+	assert(target->reg_cache->reg_list);
+	struct reg * const reg = &target->reg_cache->reg_list[regid];
+	assert(reg);
+	/* On RISC-V 0.11 targets valid value of some registers (e.g. `dcsr`)
+	 * is stored in `riscv011_info_t` itself, not in register cache. This
+	 * complicates register cache implementation.
+	 * Therefore, for now, caching registers in register cache is disabled
+	 * for all registers, except for reads of GPRs.
+	 */
+	assert(!reg->dirty);
+	int result = register_write(target, regid, value);
+	if (result != ERROR_OK)
+		return result;
+	reg_cache_set(target, regid, value);
+	/* FIXME: x0 (zero) should not be cached on writes. */
```

Comment:
It looks like the fix would be easy. Or please correct me if I am missing something.

Or, would you like to address the `FIXME` separately at a later point?

```suggestion
	if (regid != GDB_REGNO_ZERO)
		/* Avoid polluting the cache with possibly non-zero value for x0 */
		reg_cache_set(target, regid, value);
```

Reply:
Yes, I'd like for it to be fixed in a separate PR.
The reason is, the bug is not introduced by this commit. It was already there -- the fix for 013 (#906) didn't fix 0.11.
And I think it should be tested (considering that the test is quite easy).

Reply:
Ok, understood.